src/main/java/com/google/songgame/servlets/RoomServlet.java

```diff
@@ -1,25 +1,62 @@
 package com.google.songgame.servlets;
 
 import com.google.gson.Gson;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.stream.Collectors;
+import java.util.Collections;
 
 import org.apache.hc.core5.http.ParseException;
 import com.google.gson.Gson; 
-import java.util.ArrayList;
+import java.util.*;
 
 @WebServlet("/room")
 public final class RoomServlet extends HttpServlet {
 
+  private Gson gson;
+
+  @Override
+  public void init() {
+    gson = new Gson();
+  }
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get user name
-    String name = request.getParameter("user-name");
-    response.setContentType("text/html;");
-    response.getWriter().println("Name: " + name);
+    Map<String, String> userJson = readJSONFromRequest(request);
+
+    // Save player username and userId to datastore.
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("username", getValuesList(userJson).get(0));
```

Comment:
This code is somewhat brittle and prone to breaking if the order of data sent from the client changes (e.g. if the client starts sending userId first and username second).

Instead, since you already have the "userProperties" map (see my naming suggestion in a comment above), I think you can do something like userProperties.get("username") here and something similar for the userId below.

You can also see the following code for an example:
https://github.com/googleinterns/step105-2020/blob/master/src/main/java/com/google/songgame/servlets/ChatServlet.java#L58